### PR TITLE
first-commit

### PR DESCRIPTION
**Pull request by jjaykim**
TODO: inset calendar event into google calendar using API

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,131 @@
+// graphql 테스트용
```
Comment (dhmoon91) [inline .env.example:1]: why are these in `.env.example?`

File: src/entities/event.entity.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Field(() => String)
+    @Column()
+    endDateTime!: string;
+
+    @Field(() => String)
+    @Column({
+        comment: 'confirmed || cancelled',
+        default: 'confirmed',
+    })
```
Comment (dhmoon91) [inline src/entities/event.entity.ts:39]: wht do these do? 
```
 comment: 'confirmed || cancelled',
        default: 'confirmed',
```

File: src/entities/event.entity.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    status!: string;
+
+    @Field(() => Number)
+    @Column({ nullable: true })
+    userId: number;
```
Comment (dhmoon91) [inline src/entities/event.entity.ts:44]: are you sure this should be nullable?

File: src/entities/googleToken.entity.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @Column()
+    refreshToken: string;
+
+    @Field(() => Number)
+    @Column({ nullable: true })
+    userId: number;
+
+    @OneToOne(() => User, { onDelete: 'CASCADE' })
+    @JoinColumn()
+    user: User;
```
Comment (dhmoon91) [inline src/entities/googleToken.entity.ts:34]: I thought we decided to put the token in user's table?

File: src/entities/user.entity.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import {
```
Comment (dhmoon91) [inline src/entities/user.entity.ts:1]: did you rename the `user.ts` to `user.entity.ts`? 
good call.

File: src/entities/googleToken.entity.ts (hunk 4 above)
Comment (dhmoon91) [reply src/entities/googleToken.entity.ts:34]: Is there reason why you decided to put it as separate table again?

File: src/middleware/isAuth.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { MiddlewareFn } from 'type-graphql';
+
+import { MyContext } from '../types/MyContext';
+
+export const isAuth: MiddlewareFn<MyContext> = async ({ context }, next) => {
```
Comment (dhmoon91) [inline src/middleware/isAuth.ts:5]: I think name can be just `isAuthenticated`.

File: src/resolvers/UserResolver.ts [hunk 7]
```diff
@@ -1,34 +1,90 @@
...
+import { Event } from '../entities/event.entity';
+import { verifyJwtToken } from '../service/verifyJwtToken.service';
+// import { createEvent } from '../service/googleOAuth/createEvent.service';
 
 /**
  * All gql resolvers regarding User.
  */
 @Resolver()
 export class UserResolver {
+    @UseMiddleware(isAuth)
```
Comment (dhmoon91) [inline src/resolvers/UserResolver.ts:28]: nice - but why are we requiring `auth` middleware for `getUsers`?

File: src/resolvers/UserResolver.ts [hunk 8]
```diff
@@ -1,34 +1,90 @@
...
     @Query(() => [User])
     getUsers() {
-        return User.find();
+        return User.find({
+            relations: ['googleToken', 'events'],
```
Comment (dhmoon91) [inline src/resolvers/UserResolver.ts:32]: why do we need these relations? To append it to each users we find?

File: src/resolvers/UserResolver.ts [hunk 9]
```diff
@@ -42,19 +98,62 @@ export class UserResolver {
...
-            await User.update({ username }, { password: newPassword });
+            await User.update({ email }, { password: newPassword });
 
             return user;
         }
         throw new Error('Passwords do not match');
     }
+
+    @Mutation(() => Event)
+    async createEvent(
```
Comment (dhmoon91) [inline src/resolvers/UserResolver.ts:118]: this shouldn't be inside `User` resolver.

File: src/resolvers/UserResolver.ts [hunk 10]
```diff
@@ -42,19 +98,62 @@ export class UserResolver {
...
+        @Arg('data') data: EventInput,
+        @Ctx() ctx: MyContext
+    ): Promise<Event | null> {
+        try {
+            const accessToken = await verifyJwtToken(ctx.req.session.jwt!);
```
Comment (dhmoon91) [inline src/resolvers/UserResolver.ts:123]: we are now both using jwt AND session is that necessay with SSO?

File: src/server.ts [hunk 11]
```diff
@@ -4,37 +4,93 @@ import 'reflect-metadata';
...
         type: 'postgres',
         database: 'termin_dev',
         username: 'termin_admin',
         password: 'termin',
-        logging: true,
-        synchronize: false,
-        entities: [User],
+        logging: false,
+        synchronize: true,
+        entities: [`${__dirname}/entities/**/*.entity.{ts,js}`],
```
Comment (dhmoon91) [inline src/server.ts:24]: good jobb

File: src/server.ts [hunk 12]
```diff
@@ -4,37 +4,93 @@ import 'reflect-metadata';
...
+                maxAge: 1000 * 60 * 60 * 24 * 90,
+            },
+        })
+    );
+
     const schema = await buildSchema({
         // Load all resolvers
-        resolvers: [`${__dirname  }/resolvers/**/*.{ts,js}`],
+        resolvers: [`${__dirname}/resolvers/**/*.{ts,js}`],
+        authChecker: ({ context: { req } }) => {
```
Comment (dhmoon91) [inline src/server.ts:73]: why do you need this? Doesnt `isAuth` middleware handle the same thing?

File: src/server.ts [hunk 13]
```diff
@@ -4,37 +4,93 @@ import 'reflect-metadata';
...
+            if (req.session.userId) {
+                return true;
+            }
+
+            return false;
+        },
     });
 
+    // Google OAuth redirect
+    app.get('/test', googleOAuthHandler);
```
Comment (dhmoon91) [inline src/server.ts:83]: don't call it test

File: src/service/verifyJwtToken.service.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+// This service will help verifying the JWT Token of Google or Apple
+/**
+ * TODO: more subdivision of validation
+ *
+ * 1. access token and refresh token are all expired => return error
+ * 2. access toekan is expired, but refresh token is valid => re-generate the access token
+ * 3. access token is valid, but refresh token is expired => re-generate the refresh token
+ * 4. access toekn and refresh token are all valid => verified
+ * */
+
```
Comment (dhmoon91) [inline src/service/verifyJwtToken.service.ts:12]: the doc string doesn't match what this function does. 
This function is just verifying jwt

File: src/service/googleOAuth/createGoogleUser.service.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
+import { Request } from 'express';
+
+import { GoogleToken } from '../../entities/googleToken.entity';
+import { User } from '../../entities/user.entity';
+import { UserInput } from '../../types/UserInput';
+
+export const createGoogleUser = async (
```
Comment (dhmoon91) [inline src/service/googleOAuth/createGoogleUser.service.ts:7]: Doc string describing what this function does.

File: src/service/googleOAuth/googleHandler.service.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
+import { Request, Response } from 'express';
+
+import { User } from '../../entities/user.entity';
+
+import { getGoogleUser } from './getGoogleUser.service';
+import { getGoogleToken } from './getGoogleToken.service';
+import { createGoogleUser } from './createGoogleUser.service';
+import { getGoogleJwt } from './getGoogleJwt.service';
+
+export async function googleOAuthHandler(req: Request, res: Response) {
```
Comment (dhmoon91) [inline src/service/googleOAuth/googleHandler.service.ts:10]: doc string

File: src/service/googleOAuth/getGoogleUser.service.ts [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        firstName: resAPI.data.given_name,
+        lastName: resAPI.data.family_name,
+        picture: resAPI.data.picture,
+    };
+
+    return result;
+
+    // id_token을 jwt decode 하는 방법
+    /**
+  const resJWT = jwt.decode(id_token) as UserInfo;
```
Comment (dhmoon91) [inline src/service/googleOAuth/getGoogleUser.service.ts:37]: remove these comments if you don't need it

File: src/service/googleOAuth/getGoogleUser.service.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+import { UserInput } from '../../types/UserInput';
+
+interface UserInfo {
+    email: string;
+    given_name: string;
+    family_name: string;
+    picture: string;
+}
+
+export const getGoogleUser = async (
```
Comment (dhmoon91) [inline src/service/googleOAuth/getGoogleUser.service.ts:12]: doc string

**Comment by dhmoon91** [comment]
let's refcator all google related as follow: 
`service/google/` folder

`service/google/google_sso.ts`
`service/google/google_calendar.ts`
`service/google/util.ts`
and put relative functions in one file.

**Comment by dhmoon91** [comment]
Didn't you say you resolved `TODO: inset calendar event into google calendar using API` ?

File: src/service/googleOAuth/createGoogleUser.service.ts [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    req: Request
+): Promise<User> => {
+    let user: any;
+
+    // Store the values into DB
+    const refreshToken = GoogleToken.create({
+        calendarId: googleUser.email,
+        refreshToken: refreshtoken,
+    });
+    await refreshToken.save();
```
Comment (dhmoon91) [inline src/service/googleOAuth/createGoogleUser.service.ts:19]: you are calling this at the end of this anyway. No need to call it here I believe?

File: src/service/googleOAuth/createGoogleUser.service.ts (hunk 15 above)
Comment (dhmoon91) [reply src/service/googleOAuth/createGoogleUser.service.ts:7]: This is confusing too. 
You are not only creating google user here. 
You are only resetting token if user signed up using normal email and password.

Whats the core purpose of this function?

File: src/service/googleOAuth/googleHandler.service.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        const googleToken = await getGoogleToken({ code });
+
+        // get user info with tokens
+        const googleUser = await getGoogleUser(
+            googleToken.accessToken,
+            googleToken.idToken
+        );
+
+        // The user is synchronizing with Google
+        if (googleToken.refreshToken) {
```
Comment (dhmoon91) [inline src/service/googleOAuth/googleHandler.service.ts:27]: are we actually `synchronizing`? What's getting sychronized? Are we overwriting something?

**Comment by dhmoon91** [comment]
You are missing bunch of logics here and there. 

You say get something but you are creating users directly from helper function. 
Refactor it please (either change the names or separate out the creation and update parts)

File: package.json [hunk 21]
```diff
@@ -6,15 +6,20 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
     "start": "node ./dist/server.ts",
-    "dev": "nodemon ./src/server.ts",
+    "dev": "nodemon  ./src/server.ts",
```
Comment (dhmoon91) [inline package.json:9]: extra space?

File: src/entities/event.entity.ts (hunk 2 above)
Comment (jjaykim) [reply src/entities/event.entity.ts:39]: If the user canceled the reservation, I thought we should classify it for the owner

File: src/entities/googleToken.entity.ts (hunk 4 above)
Comment (jjaykim) [reply src/entities/googleToken.entity.ts:34]: 이것도 월요일에 말씀드렸듯이,
유저가 Termin 으로 회원가입하고, 구글 캘린더만 이용할때, 서로 이메일이 다른 경우가 생길수도 있기때문에 따로 entity 를 나눴습니다.

File: src/middleware/isAuth.ts (hunk 6 above)
Comment (jjaykim) [reply src/middleware/isAuth.ts:5]: changed!

File: src/resolvers/UserResolver.ts (hunk 7 above)
Comment (jjaykim) [reply src/resolvers/UserResolver.ts:28]: oh my mistake lol 
it should be at `getProfile` query.
moved it!

File: src/server.ts (hunk 12 above)
Comment (jjaykim) [reply src/server.ts:73]: 감사합니다! 지워야하는게 맞는데 까먹었습니다,,ㅎㅎ middlewae만 사용할 예정입니다!

File: src/server.ts (hunk 13 above)
Comment (jjaykim) [reply src/server.ts:83]: 네!

File: src/service/googleOAuth/getGoogleUser.service.ts (hunk 17 above)
Comment (jjaykim) [reply src/service/googleOAuth/getGoogleUser.service.ts:37]: will use JWT token

File: package.json (hunk 21 above)
Comment (jjaykim) [reply package.json:9]: mistake,,

File: src/service/verifyJwtToken.service.ts (hunk 14 above)
Comment (jjaykim) [reply src/service/verifyJwtToken.service.ts:12]: I moved this function to the `middleware` folder.
And make other functions in `service/google/utils/ts`

File: src/service/googleOAuth/createGoogleUser.service.ts (hunk 15 above)
Comment (jjaykim) [reply src/service/googleOAuth/createGoogleUser.service.ts:7]: I separated `saveRefreshToken` and `createUse`

File: src/service/googleOAuth/googleHandler.service.ts (hunk 16 above)
Comment (jjaykim) [reply src/service/googleOAuth/googleHandler.service.ts:10]: okay!

File: src/service/googleOAuth/getGoogleUser.service.ts (hunk 18 above)
Comment (jjaykim) [reply src/service/googleOAuth/getGoogleUser.service.ts:12]: okay!

File: src/service/googleOAuth/createGoogleUser.service.ts (hunk 19 above)
Comment (jjaykim) [reply src/service/googleOAuth/createGoogleUser.service.ts:19]: fixed it

File: src/service/googleOAuth/googleHandler.service.ts (hunk 20 above)
Comment (jjaykim) [reply src/service/googleOAuth/googleHandler.service.ts:27]: I wrote the wrong word. I changed!

File: src/entities/event.entity.ts (hunk 2 above)
Comment (dhmoon91) [reply src/entities/event.entity.ts:39]: can you add comment for all the fields? 
and having comment: 'confirmed||cancelled' doesn't mean anything. Be more descriptive of what this fields represent;
eg; `Event status whether canceled or confirmed`

File: src/entities/googleRefreshToken.entity.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    OneToOne,
+    JoinColumn,
+} from 'typeorm';
+import { ObjectType, Field, ID } from 'type-graphql';
+
+import { User } from './user.entity';
+
+@Entity()
+@ObjectType()
+export class GoogleRefreshToken extends BaseEntity {
```
Comment (dhmoon91) [inline src/entities/googleRefreshToken.entity.ts:15]: this table got `calendarID' and it doesn't mean anything with 'GoogleRefreshToken' table.

File: src/entities/user.entity.ts [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Field(() => String, { nullable: true })
+    @Column({ nullable: true })
+    picture?: string;
+
+    @Field(() => Boolean, { nullable: true })
+    @Column({
+        default: false,
+    })
+    hasGoogleCal: boolean;
```
Comment (dhmoon91) [inline src/entities/user.entity.ts:46]: why do you need this??

File: src/resolvers/GoogleEventResolver.ts [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+import { hasGoogleJWT } from '../middleware/hasGoogleJWT';
+
+/**
+ * All gql resolvers regarding Google Calendar.
+ */
+@Resolver()
+export class GoogleEventResolver {
+    @Mutation(() => Event)
+    @UseMiddleware(hasGoogleJWT)
+    async createEvent(
```
Comment (dhmoon91) [inline src/resolvers/GoogleEventResolver.ts:22]: leave docstring comment for this function

File: src/resolvers/GoogleEventResolver.ts [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+            return event;
+        } catch (err: any) {
+            throw new Error(`Error: ${err}`);
+        }
+    }
+
+    @Mutation(() => Event)
+    @UseMiddleware(hasGoogleJWT)
+    async cancelEvent(@Arg('eventId') eventId: string, @Ctx() ctx: MyContext) {
```
Comment (dhmoon91) [inline src/resolvers/GoogleEventResolver.ts:52]: leave docstring comment for this function

File: src/resolvers/GoogleEventResolver.ts [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+    cancelGoogleEvent,
+    updateGoogleEvent,
+} from '../service/google/calendar/utils';
+import { hasGoogleJWT } from '../middleware/hasGoogleJWT';
+
+/**
+ * All gql resolvers regarding Google Calendar.
+ */
+@Resolver()
+export class GoogleEventResolver {
```
Comment (dhmoon91) [inline src/resolvers/GoogleEventResolver.ts:19]: what is `GoogleEvent`? your comment says `Google Calendar`

File: src/resolvers/UserResolver.ts [hunk 27]
```diff
@@ -14,47 +24,120 @@ import { User } from '../entities/User';
...
-        const user = User.findOne({ where: { id } });
+        return User.findOne({
+            where: { id },
+            relations: ['googleRefreshToken', 'events'],
+        });
+    }
+
+    @Query(() => User)
+    @UseMiddleware(isAuthenticated)
+    getProfile(@Ctx() ctx: MyContext) {
```
Comment (dhmoon91) [inline src/resolvers/UserResolver.ts:42]: remove this function

File: src/service/google/calendar/utils.ts [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+        return tokens.data.access_token;
+    } catch (err: any) {
+        throw new Error(
+            'Failed to refresh the access token...\n Please try again'
+        );
+    }
+};
+
+/**
+ * It guarantees connection with Google
```
Comment (dhmoon91) [inline src/service/google/calendar/utils.ts:42]: `It guarantees connection with Google` how?

File: src/service/google/calendar/utils.ts [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+    } catch (err: any) {
+        throw new Error(err);
+    }
+};
+
+/**
+ * It creates the event in the user's primary Google calendar
+ * It also returns created Google event ID which is a unique key
+ */
+export const createEventGoogle = async (
```
Comment (dhmoon91) [inline src/service/google/calendar/utils.ts:80]: why do you have these in `utils` file? Theses seems like all core functionalities regarding `google_calendar` and should be put there

File: src/service/google/calendar/utils.ts [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+        });
+    } catch (err: any) {
+        throw new Error(`Error: ${err}`);
+    }
+};
+
+/**
+ * It updates the event on Google Calendar
+ */
+export const updateGoogleEvent = async (
```
Comment (dhmoon91) [inline src/service/google/calendar/utils.ts:162]: updateGoogleEvent -> updateGoogleCalendar

File: src/types/EventInput.ts [hunk 31]
```diff
@@ -0,0 +1,24 @@
...
+    description!: string;
+
+    @Field({ nullable: false })
+    startDateTime!: string;
+
+    @Field({ nullable: false })
+    endDateTime!: string;
+
+    @Field({ nullable: true })
+    googleEventId: string;
```
Comment (dhmoon91) [inline src/types/EventInput.ts:23]: you need this as input? 
Don't we just create this for the user?

**Comment by dhmoon91** [comment]
overall looks good but you are naming keeps mixing `Event` and `Google Calendar`.
We prob need more refactor later

File: src/service/google/calendar/utils.ts [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+
+import { createGoolgeJWT } from '../sso/utils';
+import { Event } from '../../../entities/event.entity';
+import { EventInput } from '../../../types/EventInput';
+import { GoogleRefreshToken } from '../../../entities/googleRefreshToken.entity';
+import { GoogleJwtPayload } from '../../../types/Google';
+import { verifyJwtToken } from '../../verifyJwtToken';
+
+/**
+ * Refreshing Google access token
```
Comment (dhmoon91) [inline src/service/google/calendar/utils.ts:14]: why do we need to refresh google access token? 
Give some details in your comments

File: src/entities/user.entity.ts (hunk 23 above)
Comment (jjaykim) [reply src/entities/user.entity.ts:46]: 유저가 예약을 변경하거나 취소할때, 
이 유저가 `Google` 을 연결했는지, `Apple`을 연결했는지를 체크하려고 했는데, 
아예 `googleRefreshToken` Entity를 수정해서 거기에 추가하여 refector 할 계획입니다.

`googleRefreshToken` entity -> `googleSSO` entity
`calendarId` column -> `userEmail` column
add `useCalendar` column 
나머진 그대로 사용 예정입니다.

File: src/resolvers/GoogleEventResolver.ts (hunk 24 above)
Comment (jjaykim) [reply src/resolvers/GoogleEventResolver.ts:22]: Okay!

File: src/resolvers/GoogleEventResolver.ts (hunk 25 above)
Comment (jjaykim) [reply src/resolvers/GoogleEventResolver.ts:52]: Okay!

File: src/resolvers/GoogleEventResolver.ts (hunk 26 above)
Comment (jjaykim) [reply src/resolvers/GoogleEventResolver.ts:19]: `GoogleCalendarResolver` 가 더 맞는 것 같습ㄴㅣ다

File: src/resolvers/UserResolver.ts (hunk 27 above)
Comment (jjaykim) [reply src/resolvers/UserResolver.ts:42]: okay!

File: src/service/google/calendar/utils.ts (hunk 28 above)
Comment (jjaykim) [reply src/service/google/calendar/utils.ts:42]: I will add docstring more detail

File: src/service/google/calendar/utils.ts (hunk 29 above)
Comment (jjaykim) [reply src/service/google/calendar/utils.ts:80]: `connectGoogleCal` in `google_calendar.ts` file 이 유저가 calendar를 사용하겠다고 했을때 다시 SSO 로그인을 통해서 refresh token을 가져오기때문에 가장 중요한 함수라고 생각햇습니다.
Refresh token을 다시 가져오는 이유는 
맨처음 google SSO 로그인만 했을경우에는, 오직 로그인을 위한 refresh token을 google한테 받기 때문에, calendar에 접근이 불가능합니다. (scope가 다르기때문에)
그래서 calendar를 사용하기 위해서 다시 token을 받아야합니다.

event를 추가하고, 수정하고, 삭제하는 기능들은 캘린더를 연결한 뒤에 사용하는 additional functionalities 라고 생각해서 `utils`에 넣었습니다.

File: src/service/google/calendar/utils.ts (hunk 30 above)
Comment (jjaykim) [reply src/service/google/calendar/utils.ts:162]: okay!

File: src/types/EventInput.ts (hunk 31 above)
Comment (jjaykim) [reply src/types/EventInput.ts:23]: 나중에 evnet(예약)을 변경하거나 취소했을때 그 eventId가 필요합니다. 
`googleEventId`라고 한 이유는 나중에 `apple` 캘린더까지 사용했을 경우를 대비해서 정했습니다.

Comment (jjaykim) [reply src/types/EventInput.ts:23]: column comment에 자세하게 남기겠습니다!

File: src/service/google/calendar/utils.ts (hunk 32 above)
Comment (jjaykim) [reply src/service/google/calendar/utils.ts:14]: okay!

File: src/resolvers/UserResolver.ts (hunk 27 above)
Comment (jjaykim) [reply src/resolvers/UserResolver.ts:42]: 아직 `google sso` 테스트를 위해서 필요해서 나중에 지우겠습니다,,ㅠ

**Pull request merged**